src/distributions/bernoulli.rs

```diff
@@ -0,0 +1,94 @@
+//! The Bernoulli distribution.
+
+use Rng;
+use distributions::Distribution;
+
+/// The Bernoulli distribution.
+///
+/// This is a special case of the Binomial distribution where `n = 1`.
+///
+/// # Example
+///
+/// ```rust
+/// use rand::distributions::{Bernoulli, Distribution};
+///
+/// let d = Bernoulli::new(0.3);
+/// let v = d.sample(&mut rand::thread_rng());
+/// println!("{} is from a Bernoulli distribution", v);
+/// ```
+#[derive(Clone, Copy, Debug)]
+pub struct Bernoulli {
+    /// Probability of success, relative to the maximal integer.
+    p_int: u64,
+}
+
+impl Bernoulli {
+    /// Construct a new `Bernoulli` with the given probability of success `p`.
+    ///
+    /// # Panics
+    ///
+    /// If `p < 0` or `p > 1`.
+    ///
+    /// # Precision
+    ///
+    /// For `p = 1.0`, the resulting distribution will always generate true.
+    /// For `p = 0.0`, there is a chance of 2<sup>64</sup> to incorrectly
+    /// generate true.
+    #[inline]
+    pub fn new(p: f64) -> Bernoulli {
+        assert!(p >= 0.0, "Bernoulli::new called with p < 0");
+        assert!(p <= 1.0, "Bernoulli::new called with p > 1");
+        let p_int = if p != 1.0 {
+            (p * (::core::u64::MAX as f64)) as u64
```

Comment:
I *guess* what happens here is that `u64::MAX` gets cast to `2.0.powi(64)`, in which case the multiplication is probably only an adjustment to the exponent of `p`. So if we stick with this method, IMO it makes more sense to test `r < self.p_int` and document that `Bernoulli::new(1.0)` may sample incorrectly (instead of 0.0), because then it's *only* 1.0 and small values not a multiple of `2^-64` which aren't sampled exactly.

Reply:
Consider 0.5: this should get mapped exactly to `2^63`; exactly half of all possible `u64` values are less than this.

This is only the case because `u64::MAX as f64` isn't representable and essentially becomes `2^64`.

Reply:
I think it is fine to have an error smaller than 2<sup>-64</sup>. If someone wants more precision, they should probably just use `Uniform` with only integers.